File: dmoj/ccc23s3/ccc23s3.cpp
#include <bits/stdc++.h>
using namespace std;
#define FR(i, n) for(int i = 0; i < (n); i++)

const int MN = 2005;
int n, m, r, c;
bool swapped;
char grid[MN][MN];

bool solve(){
    memset(grid, 'a', sizeof(grid));
    if(n == r && m == c) return true;
    if(m == c) swap(n, m), swap(r, c), swapped = true;

    if(n != r){
        FR(i, n-r) grid[i][0] = 'b';
        FR(i, m-c) grid[0][i] = 'b';
        grid[0][0] = 'c';
    }
    else {
        FR(i, (m-c)/2) grid[0][i] = grid[0][m-i-1] = 'b';
        if((m-c) & 1){
            if(m & 1) grid[0][m/2] = 'b';
            else return false;
        }
    }
    return true;
}

int main(){
    scanf("%d %d %d %d", &n, &m, &r, &c);
    if(!solve()) return !printf("IMPOSSIBLE\n");
    if(swapped){
        swap(n, m);
        FR(i, max(n, m)) FR(j, i) swap(grid[i][j], grid[j][i]);
    }
    FR(i, n) grid[i][m] = 0, puts(grid[i]);
}

File: dmoj/ccc23s3/ccc23s3.py
n, m, r, c = map(int, input().split())

swapped = False
if m == c:
    n, m = m, n
    r, c = c, r
    swapped = True

grid = [['a']*m for i in range(n)]

def solve():
    if n == r and m == c: return True

    if n != r:
        for i in range(n-r): grid[i][0] = 'b'
        for i in range(m-c): grid[0][i] = 'b'
        grid[0][0] = 'c'
    else:
        for i in range((m-c)//2): grid[0][i] = grid[0][m-i-1] = 'b'
        if (m-c) % 2 == 1:
            if m % 2 == 1: grid[0][m//2] = 'b'
            else: return False
    return True

if not solve(): print("IMPOSSIBLE")
elif swapped:
    for i in range(m):
        for j in range(n):
            print(grid[j][i], end='')
        print()
else:
    for i in range(n):
        for j in range(m):
            print(grid[i][j], end='')
        print()
